src/stream.rs

```diff
@@ -435,13 +435,21 @@ pub(crate) fn connect_host(
 
             let mut proxy_response = Vec::new();
 
-            loop {
-                let mut buf = vec![0; 256];
-                let total = stream.read(&mut buf)?;
-                proxy_response.append(&mut buf);
-                if total < 256 {
+            let mut buf = [0u8; 1];
+            while let Ok(total) = stream.read(&mut buf) {
+                if total == 0 {
+                    // no more data
                     break;
                 }
+                proxy_response.push(buf[0]);
+                match buf[0] {
+                    b'\n' => {
+                        if proxy_response.ends_with(b"\r\n\r\n".as_slice()) {
```

Comment:
`ends_with` is a quite expensive call, however it's mitigated by that we only do it when we encounter an `\n`, so on balance probably ok.

I don't think `.as_slice()` does anything for you here.

Reply:
According to the docs for Vec [0] and the source [1] this is a simple byte comparison of a subslice.
This shouldn't be that expensive.

[0] https://doc.rust-lang.org/std/vec/struct.Vec.html#method.ends_with
[1] https://doc.rust-lang.org/src/core/slice/mod.rs.html#2303-2305